docker/cdash.docker

```diff
@@ -0,0 +1,117 @@
+FROM php:7.1-apache
+LABEL maintainer="Kitware, Inc. <cdash@public.kitware.com>"
+
+ARG CDASH_DB_HOST=localhost
+
+# Designates as dev build, adds testing infrastructure, et al.
+ARG DEVELOPMENT_BUILD
+
+# make sure it's set in the ENV for reference in docker-entrypoint.sh
+# TODO: rename CDASH_DEVELOPMENT_BUILD ?
+ENV DEVELOPMENT_BUILD=$DEVELOPMENT_BUILD
+
+# NODE_ENV expands to nothing if DEVELOPMENT_BUILD is unset or empty
+ENV NODE_ENV=${DEVELOPMENT_BUILD:+'development'}
+
+# NODE_ENV expands to production if NODE_ENV is unset or empty,
+# i.e. if DEVELOMPMENT_BUILD is empty the node environment is production
+ENV NODE_ENV=${NODE_ENV:-'production'}
+
+# PHP_ENV expands to empty if DEVELOPMENT_BUILD is unset or empty
+ENV PHP_ENV=${DEVELOPMENT_BUILD:+' '}
+
+# PHP_ENV expands to --no-dev (composer flag) if PHP_ENV is empty,
+# i.e. composer will not try to load require-dev packages, e.g. those
+# requiring xdebug
+ENV PHP_ENV=${PHP_ENV:-'--no-dev'}
+
+RUN apt-get update                                                         \
+ && apt-get install -y gnupg                                               \
+ && curl -sL https://deb.nodesource.com/setup_6.x | bash                   \
+ && apt-get install -y git libbz2-dev libfreetype6-dev libjpeg62-turbo-dev \
+    libmcrypt-dev libpng-dev libpq-dev libxslt-dev libxss1 nodejs unzip    \
+    wget zip                                                               \
+ && docker-php-ext-configure pgsql --with-pgsql=/usr/local/pgsql           \
+ && docker-php-ext-configure gd --with-freetype-dir=/usr/include/          \
+                                --with-jpeg-dir=/usr/include/              \
+ && docker-php-ext-install -j$(nproc) bcmath bz2 gd                        \
+    pdo_mysql pdo_pgsql xsl                                                \
+ && wget -q -O checksum https://composer.github.io/installer.sha384sum     \
+ && wget -q -O composer-setup.php https://getcomposer.org/installer        \
+ && sha384sum -c checksum                                                  \
+ && php composer-setup.php                                                 \
+    --install-dir=/usr/local/bin --filename=composer                       \
+ && php -r "unlink('composer-setup.php');"                                 \
+ && composer self-update --no-interaction
+
+# Creates the kitware home directory with directory for cdash
+WORKDIR /home/kitware/cdash
+
+# Copy CDash (current folder) into /home/kitware/cdash to set its owner
+# to www-data
+ADD . ./
+
+RUN touch "storage/logs/laravel-$(date +%F).log"
```

Comment:
Speaking of Dockerfile best practices, this line poses a problem with cache invalidation, not to mention that it interacts with the Docker cache in a way that you probably didn't expect.

The entire `RUN` line is considered when determining whether to invalidate the Docker cache (*before* running the command).  The `date` subcommand expansion is not considered, so the actual timestamp in the final image won't be just the last time you built the image, but the last time you built the image with an invalid cache (any subsequent builds that hit the cache at this point will reuse the old timestamp).

Even doing the alternative is problematic, because you would need to *unconditionally* invalidate the cache at this line, defeating the purpose of all your latter efforts to use the cache.

If something like this *needs* to be done ahead of time, I would suggest delaying it until container run time, and just make it one of the first things you do in your `ENTRYPOINT`.